panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,688 @@
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+schedule_save_tariff (CcUpdatesPanel *self)
+{
+  g_debug ("Scheduling save tariff");
+
+  if (self->save_tariff_timeout_id > 0)
+    g_source_remove (self->save_tariff_timeout_id);
+
+  self->save_tariff_timeout_id = g_timeout_add_seconds (2, save_tariff_cb, self);
+}
+
+
+/*
+ * Callbacks
+ */
+
+static void
+on_automatic_updates_switch_changed_cb (GtkSwitch      *sw,
+                                        GParamSpec     *pspec,
+                                        CcUpdatesPanel *self)
+{
+  NMConnection *connection;
+  gboolean active;
+
+  active = gtk_switch_get_active (sw);
+  get_active_connection_and_device (self, NULL, &connection, NULL);
+
+  set_automatic_updates_setting (connection, active);
+
+  /* Disable scheduled updates if automatic updates are disabled too */
+  if (!active)
+    gtk_switch_set_active (GTK_SWITCH (self->scheduled_updates_switch), FALSE);
+}
+
+static gboolean
+on_change_network_link_activated_cb (GtkLabel       *label,
+                                     gchar          *uri,
+                                     CcUpdatesPanel *self)
+{
+  g_autoptr(GError) error = NULL;
+  NMDevice *device;
+  CcShell *shell;
+
+  shell = cc_panel_get_shell (CC_PANEL (self));
+
+  get_active_connection_and_device (self, &device, NULL, NULL);
+
+  if (!device || !NM_IS_DEVICE_WIFI (device))
+    cc_shell_set_active_panel_from_id (shell, "network", NULL, &error);
+  else
+    cc_shell_set_active_panel_from_id (shell, "wifi", NULL, &error);
+
+  if (error)
+    {
+      g_warning ("Error activating panel: %s", error->message);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+on_network_changed_cb (CcUpdatesPanel *self)
+{
+  g_debug ("NetworkManager changed state");
+
+  update_active_network (self);
+}
+
+static void
+on_scheduled_updates_switch_changed_cb (GtkSwitch      *sw,
+                                        GParamSpec     *pspec,
+                                        CcUpdatesPanel *self)
+{
+  g_debug ("Scheduled Updates changed state");
+
+  schedule_save_tariff (self);
+}
+
+static void
+on_tariff_changed_cb (CcTariffEditor *tariff_editor,
+                      CcUpdatesPanel *self)
+{
+  g_debug ("The saved tariff changed");
+
+  schedule_save_tariff (self);
+}
+
+
+/*
+ * GObject overrides
+ */
+
+static void
+cc_updates_panel_dispose (GObject *object)
+{
+  CcUpdatesPanel *self = (CcUpdatesPanel *)object;
+
+  if (self->save_tariff_timeout_id > 0)
+    {
+      save_tariff (self);
+      g_source_remove (self->save_tariff_timeout_id);
+      self->save_tariff_timeout_id = 0;
+    }
+
+  G_OBJECT_CLASS (cc_updates_panel_parent_class)->dispose (object);
+}
+
+static void
+cc_updates_panel_finalize (GObject *object)
+{
+  CcUpdatesPanel *self = (CcUpdatesPanel *)object;
+
+  g_cancellable_cancel (self->cancellable);
+  g_clear_object (&self->cancellable);
+  g_clear_object (&self->current_device);
+  g_clear_object (&self->nm_client);
+
```

Comment:
Need to disconnect `changed_id` too.

Reply:
At first, I did that. But appearently NetworkManager destroys all the `NMDevices` and we end up with an invalid source id at this point. Disconnecting in `dispose()` also yields the same results.

Reply:
That doesn’t make sense, since you’re holding a strong reference to the device in `current_device`, right? As far as I understand it, `changed_id` is a signal handler ID, not a source ID, though. Were you using `g_source_remove()` here? That wouldn’t be valid; use `g_signal_handler_disconnect()`.

Reply:
/me facepalms
Please pretend we never had this discussion.